### ISSUE TITLE
FE | Identity Team | Direct Deposit | Update CTA Widget Sign-in Behavior

### ISSUE THREAD

**Issue by andaleliz** [department-of-veterans-affairs/va.gov-team#59438]
## Next Steps 
Identity team to pick this up 

## Background

On the change direct deposit static page ([staging link](https://staging.va.gov/change-direct-deposit/)), we have a sign in CTA that currently sends people back to the static page, instead of taking them directly to the tool when they sign in. To get people more efficiently to the tool, let's guide directly to `/profile/direct-deposit` after signing in, so they don't have to click the link again on the static page.

Here's a video of the current behavior.

https://github.com/department-of-veterans-affairs/va.gov-team/assets/55992116/c730e069-7891-4584-8251-1d7641e09423

This is part of the effort to enhance the logged-in experience; @mnorthuis called this out as a good initial opportunity that could lead to improve the UX in many places across the site. Many tools use this sign in CTA and it behaves differently depending on where it's implemented (see below) Since this page falls under the purview of our team, and we worked on the research, it makes sense for us to take this on.

### Consideration
On other static pages with this widget, there's a weird flash of the static page before the person is directed to the tool. Can we avoid that? [Here's a Slack thread with more info and examples.](https://dsva.slack.com/archives/C52CL1PKQ/p1685556303394659)

## Tasks
- [x] Did discovery work around what the solution needs to be
- [x] Worked out a solution, and identity team needs an similar ticket created for them to address this, since they own the code where the rerouting would happen post-login.
- [x] Ticket created
  

## Acceptance Criteria
- [x] Ticket created for identity team

## Validation
- Who can validate this ticket? anyone on the team

**Comment by andaleliz** [comment]
This problem has been around for a while; I've asked for any previous documentation around why this happens or attempts to fix it. Will add more details to the issue description as I get them.

**Comment by adamwhitlock1** [comment]
@andaleliz I am having trouble reproducing the desired behavior.

When I log in via the button on `/change-direct-deposit/` I land back on that same page/url and have to click the button within the CTA once more to actually navigate to the `profile/direct-deposit` page of our app. It doesn't seem to redirect at all for me.

I am able to see the undesired behavior on the other pages that you linked as example of the page flashing and then redirecting. I will continue to look at how that is working, but it seems to boil down to the login flow not doing the 'redirecting', and instead the redirecting is done by the CTA widget itself in a series of steps like: 

loading -> checking login status -> checking if it should redirect somewhere -> actually doing the redirect action

The code that configures these CTAs whether to redirect or do other things lives here:

https://github.com/department-of-veterans-affairs/vets-website/blob/main/src/applications/static-pages/cta-widget/ctaWidgets.js#L178-L189

^ I highlighted specifically the CTA code that is used on https://www.va.gov/housing-assistance/home-loans/check-coe-status which has the redirect set to `true`

**Comment by adamwhitlock1** [comment]
Ahhhh I see, thanks for the clarification! 

If we did want to prevent the flash of page content when a redirect is set to true, then I think the cta would need to be adjusted to somehow display a loading indicator that hides the _entire_ existing page until the redirect logic has been executed. Either that, or maybe figuring out how the redirect is set up to be part of the login process like the My VA redirect from the homepage login.

I'll keep poking at this though now that I know more clearly what we are looking for.

**Comment by andaleliz** [comment]
You're welcome!

Both of those are great ideas and an improvement over that weird flash of content, or having to hit the button again.  IMO this one would be ideal:

>maybe figuring out how the redirect is set up to be part of the login process like the My VA redirect from the homepage login.

I'm interested to see what you learn now that we've clarified things! Thanks for working on this Adam 🕵️

**Comment by adamwhitlock1** [comment]
I dug into the ctas that live within the 'static-pages' application where the direct deposit cta lives, and after evaluating the component and the related logic, I think the limitation instead lives in the authentication components that would be part of the identity team's code.

When the login modal shows up and you click on a particular login method (id.me dslogon etc) that is when the current url is captured and stored in session storage, which can then be used to redirect back to that page once the authentication is done.

The code lives within the AuthApp and related utilities, and doesn't have a way to specify a different 'custom' redirect url:

https://github.com/department-of-veterans-affairs/vets-website/blob/main/src/platform/user/authentication/utilities.js#L142

There is the basic pattern matching that is done for redirecting the login from the homepage to my-va that is somewhat closer to what we would want, but that would not quite fit our needs in the cta either.

So overall the conclusion I have come to:

We would need to open an issue with the Identity team to determine the LOE involved in being able to specify a custom redirect url for post login that we could use in the ctas so it would redirect to our desired location.

**Comment by asg5704** [comment]
I think you are right @adamwhitlock1 possibly being an authentication thing. I know we helped make the 'custom' redirect for the Profile (auth-exp) team with the redirect to `/my-va` but I think we want to find a better way to do the redirect.

I have an idea on how I'd like to fix this issue and the LOE seems like a sprint (maybe 2). Can you add the `identity` label and I'll prioritize with my team to see when we can get this work in.

**Comment by asg5704** [comment]
Quick Proof of Concept to showcase point

https://github.com/department-of-veterans-affairs/va.gov-team/assets/67602137/b06bd1e3-68d2-4674-9b3e-0399d0841599

**Comment by adamwhitlock1** [comment]
Thanks @asg5704 for the quick response and awesome POC! That is exactly what I was basically thinking.

I added the identity label to this, and if yall can work this into the next couple sprints that would be great.

**Comment by adamwhitlock1** [comment]
@mtcA6 Since the identity team will be looking into this work, I would suggest maybe we can duplicate this ticket, and send it over there for their team to work in when appropriate, and then we can close this ticket out on our side.

**Comment by asg5704** [comment]
> @mtcA6 Since the identity team will be looking into this work, I would suggest maybe we can duplicate this ticket, and send it over there for their team to work in when appropriate, and then we can close this ticket out on our side.

Yes please 🙏

**Comment by andaleliz** [comment]
What's the ticket number for the new ticket the Identity team is working on to complete this? I'd like to follow along please :)

**Comment by adamwhitlock1** [comment]
@andaleliz This work was actually already completed and is live. The [PR where the change was made](https://github.com/department-of-veterans-affairs/vets-website/pull/24756) was merged before the weekend.

I checked it out on the direct deposit page within staging (i used user 202 as I knew they had direct deposit info set up) and it redirected immediately without any funky flash!

**Comment by andaleliz** [comment]
YAY! That's so cool.  I notice the funky flash also doesn't happen on the Claim Status Tool page so it seems like it fixed things in a couple of different places.

I also checked the sign in CTA widget on for the payment history tool (va.gov/va-payment-history/) and that one doesn't redirect. So, if I understand correctly, it seems like teams will need to be sure their sign-in CTA widget has that redirect boolean set to true. Is that right? Is there a process for letting people know they can do that free of wonky-redirect fears?

**Comment by asg5704** [comment]
@andaleliz Yes, that is correct. @adamwhitlock1 brought up a good point that it could impact Team Sites and we linked up to make the determination based on how the Direct Deposit and other `redirect: true` apps panned out. So for teams who want the same functionality all they have to do is set their `redirect: true` and the rest is taken care of 😄

**Comment by andaleliz** [comment]
@asg5704 thank you so much! This is exciting, I know our content/IA folks have been looking forward to getting this fixed for some time.  Thanks for the info! I'll figure out how to spread the word 📣
